pkg/maglev/maglev.go

```diff
@@ -89,90 +150,156 @@ func getPermutation(backends []string, m uint64, numCPU int) []uint64 {
+// sorting and for hashing. To make sure the representation stays stable,
+// this method is reproducing parts of Backend.String(), AddrCluster.String(),
+// etc.
+//
+// This MUST NOT be changed as otherwise we may end up with different maglev
+// lookup tables on different nodes during upgrades. To introduce a change to
+// this we would need to add a new flag to switch to new algorithm to migrate
+// new installs over.
+func (bi *BackendInfo) setHashString() {
+	if bi.hashString != "" {
+		return
+	}
 
-	return permutation[:bCount*int(m)]
+	var b strings.Builder
+	b.WriteByte('[')
+	a := bi.Addr
+	if a.IsIPv6() {
+		b.WriteByte('[')
+		b.WriteString(a.AddrCluster.String())
+		b.WriteString("]:")
+	} else {
+		b.WriteString(a.AddrCluster.String())
+		b.WriteByte(':')
+	}
+	b.WriteString(strconv.FormatUint(uint64(a.Port), 10))
+	b.WriteByte('/')
+	b.WriteString(a.Protocol)
+	if a.Scope == loadbalancer.ScopeInternal {
+		b.WriteString("/i")
+	}
+	b.WriteString(",State:active]")
+	bi.hashString = b.String()
 }
 
-// GetLookupTable returns the Maglev lookup table of the size "m" for the given
-// backends. The lookup table contains the IDs of the given backends.
+// GetLookupTable returns the Maglev lookup table for the given backends.
+// The lookup table contains the IDs of the given backends.
 //
 // Maglev algorithm might produce different lookup table for the same
 // set of backends listed in a different order. To avoid that sort
-// backends by name, as the names are the same on all nodes (in opposite
+// backends by the hash, as these are the same on all nodes (in opposite
 // to backend IDs which are node-local).
 //
 // The weights implementation is inspired by https://github.com/envoyproxy/envoy/pull/2982.
 //
 // A backend weight is honored by altering the frequency how often a backend's turn is
 // selected.
 // A backend weight is multiplied in each turn by (n + 1) and compared to
-// weightCntr[backendName] value which is an incrementation of weightSum (but starts at
+// weightCntr[index] value which is an incrementation of weightSum (but starts at
 // backend's weight / number of backends, so that each backend is selected at least once). If this is lower
-// than weightCntr[backendName], another backend has a turn (and weightCntr[backendName]
+// than weightCntr[index], another backend has a turn (and weightCntr[index]
 // is incremented). This way we honor the weights.
-func GetLookupTable(backendsMap map[string]*loadbalancer.Backend, m uint64) []int {
-	if len(backendsMap) == 0 {
-		return nil
+func (ml *Maglev) GetLookupTable(backends iter.Seq[BackendInfo]) []loadbalancer.BackendID {
+	ml.mu.Lock()
+	defer ml.mu.Unlock()
+
+	var backendInfos []BackendInfo
+	if ml.backendInfosBuffer != nil {
+		// Reuse previous buffer.
+		backendInfos = ml.backendInfosBuffer[:0]
+	}
+	for b := range backends {
+		b.setHashString()
+		backendInfos = append(backendInfos, b)
 	}
+	ml.backendInfosBuffer = backendInfos[:0]
```

Comment:
Sure, we can do it that way. Changed the code to just use `ml.backendInfosBuffer` everywhere. I don't see the point of calculating the size since we reuse the buffer: we will quickly find a suitable sized buffer to use and it doesn't matter so much that the first few times we realloc a bit and that's cheaper than always iterating twice.